File: backend/src/main/java/com/hgw/gestionale/auth/service/AuthService.java
package com.hgw.gestionale.auth.service;

import com.hgw.gestionale.auth.dto.LoginRequest;
import com.hgw.gestionale.auth.dto.LoginResponse;
import com.hgw.gestionale.auth.dto.RegisterRequest;
import com.hgw.gestionale.auth.entity.RefreshToken;
import com.hgw.gestionale.auth.repository.RefreshTokenRepository;
import com.hgw.gestionale.security.JwtService;
import com.hgw.gestionale.user.entity.Role;
import com.hgw.gestionale.user.entity.User;
import com.hgw.gestionale.user.repository.UserRepository;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseCookie;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import org.springframework.http.HttpStatus;
import org.springframework.web.server.ResponseStatusException;

import java.time.LocalDateTime;

@Service
@RequiredArgsConstructor
public class AuthService {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final AuthenticationManager authenticationManager;
    private final JwtService jwtService;
    private final RefreshTokenRepository refreshTokenRepository;

    @Value("${auth.refresh-token.expiration-days}")
    private int refreshTokenExpirationDays;

    @Value("${auth.refresh-token.expiration-days-remember-me}")
    private int refreshTokenExpirationDaysRememberMe;

    @Value("${auth.refresh-token.cookie-secure}")
    private boolean cookieSecure;

    public void register(RegisterRequest request) {
        User user = User.builder()
                .username(request.username())
                .passwordHash(passwordEncoder.encode(request.password()))
                .role(Role.COMMESSO)
                .enabled(true)
                .build();
        userRepository.save(user);
    }

    public LoginResponse login(LoginRequest request, HttpServletResponse response) {
        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        request.username(),
                        request.password()
                )
        );

        User user = userRepository.findByUsername(request.username())
                .orElseThrow(() -> new RuntimeException("User not found"));

        String accessToken = jwtService.generateAccessToken(user.getUsername(), user.getRole().name());
        String refreshTokenValue = jwtService.generateRefreshToken();
        boolean rememberMe = Boolean.TRUE.equals(request.rememberMe());

        LocalDateTime expiresAt = rememberMe
                ? LocalDateTime.now().plusDays(30)
                : LocalDateTime.now().plusDays(1);

        RefreshToken refreshToken = RefreshToken.builder()
                .token(refreshTokenValue)
                .user(user)
                .expiresAt(expiresAt)
                .revoked(false)
                .build();

        refreshTokenRepository.save(refreshToken);

        ResponseCookie cookie = ResponseCookie.from("refresh_token", refreshTokenValue)
                .httpOnly(true)
                .secure(cookieSecure)
                .path("/api/auth")
                .maxAge(rememberMe ? 60L * 60 * 24 * refreshTokenExpirationDaysRememberMe : 60L * 60 * 24 * refreshTokenExpirationDays)
                .sameSite("Lax")
                .build();

        response.addHeader("Set-Cookie", cookie.toString());

        return new LoginResponse(accessToken);
    }

    public LoginResponse refresh(HttpServletRequest request, HttpServletResponse response) {

        String refreshTokenValue = extractRefreshTokenFromCookies(request);

        RefreshToken refreshToken = refreshTokenRepository.findByToken(refreshTokenValue)
                .orElseThrow(() -> {
                        removeTokenFromCookies(response);
                        throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Refresh token not found");
                });

        if (refreshToken.isRevoked() || refreshToken.getExpiresAt().isBefore(LocalDateTime.now())) {
                logout(request, response);
                throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Refresh token expired or revoked");
        }

        String accessToken = jwtService.generateAccessToken(
                refreshToken.getUser().getUsername(),
                refreshToken.getUser().getRole().name()
        );

        return new LoginResponse(accessToken);
    }

    private String extractRefreshTokenFromCookies(HttpServletRequest request) {
        if (request.getCookies() == null) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "No cookies");
        }

        for (Cookie cookie : request.getCookies()) {
            if ("refresh_token".equals(cookie.getName())) {
                return cookie.getValue();
            }
        }

        throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Refresh token cookie not found");
    }

    public void logout(HttpServletRequest request, HttpServletResponse response) {

        String refreshTokenValue = extractRefreshTokenFromCookies(request);

        refreshTokenRepository.findByToken(refreshTokenValue)
                .ifPresent(refreshTokenRepository::delete);

        removeTokenFromCookies(response);
    }

    private void removeTokenFromCookies(HttpServletResponse response){
        ResponseCookie cookie = ResponseCookie.from("refresh_token", "")
                .httpOnly(true)
                .secure(cookieSecure)
                .path("/api/auth")
                .maxAge(0)
                .sameSite("Lax")
                .build();

        response.addHeader("Set-Cookie", cookie.toString());
    }
}

File: backend-node/src/service/auth.service.ts
import bcrypt from "bcryptjs";
import { LoginRequest, LoginResponse, RegisterRequest } from "../dto/auth.dto";
import { db } from "../config/db";
import { refreshTokens, users } from "../schema";
import { eq } from "drizzle-orm";
import { HttpStatus } from "../common/httpStatus";
import { HttpError } from "../common/httpError";
import { generateAccessToken, generateRefreshToken, extractRefreshTokenFromCookies, clearRefreshTokenCookie } from "../common/token";
import { Request, Response } from "express";
import { cookieConfig } from "../config/cookie";

export class AuthService {
    async register(request: RegisterRequest): Promise<void> {
        const passwordHash = await bcrypt.hash(request.password, 10);

        await db.insert(users).values({
            username: request.username,
            nome: request.nome,
            passwordHash,
            email: request.email,
            role: request.role,
            enabled: true
        })
    }

    async login(request: LoginRequest, res: Response): Promise<LoginResponse> {
        const result = await db.select().from(users).where(eq(users.username, request.username));
        if (!result.length) throw new HttpError(HttpStatus.UNAUTHORIZED, "Credenziali non valide");

        const utente = result.at(0)!;

        const passwordValida = await bcrypt.compare(request.password, utente.passwordHash);
        if (!passwordValida) throw new HttpError(HttpStatus.UNAUTHORIZED, "Credenziali non valide");

        const accessToken: string = generateAccessToken({ sub: utente.username, role: utente.role });
        const refreshToken: string = generateRefreshToken();
        const rememberMe: boolean = request.rememberMe;

        const days = rememberMe ? cookieConfig.rememberMeDays : cookieConfig.defaultDays;

        const expiresAt = new Date();
        expiresAt.setDate(expiresAt.getDate() + days);

        await db.insert(refreshTokens).values({
            userId: utente.id,
            token: refreshToken,
            expiresAt,
            revoked: false
        })

        res.cookie("refreshToken", refreshToken, {
            httpOnly: true,
            secure: cookieConfig.secure,
            path: "/api/auth",
            sameSite: cookieConfig.sameSite,
            maxAge: days * 24 * 60 * 60 * 1000
        });

        return { accessToken };
    }

    async logout(req: Request, res: Response): Promise<void> {
        const refreshTokenValue = extractRefreshTokenFromCookies(req);

        const result = await db.select().from(refreshTokens).where(eq(refreshTokens.token, refreshTokenValue));
        const token = result.at(0);

        if (!token || token.revoked) {
            clearRefreshTokenCookie(res);
            return;
        }

        await db.update(refreshTokens)
            .set({ revoked: true })
            .where(eq(refreshTokens.token, refreshTokenValue));

        clearRefreshTokenCookie(res);
    }

    async refresh(req: Request, res: Response): Promise<LoginResponse> {
        const refreshTokenValue = extractRefreshTokenFromCookies(req);
        const tokenResult = await db.select().from(refreshTokens).where(eq(refreshTokens.token, refreshTokenValue));
        const refreshToken = tokenResult.at(0);

        if (!refreshToken) {
            clearRefreshTokenCookie(res);
            throw new HttpError(HttpStatus.UNAUTHORIZED, "Refresh token not found");
        }

        if (refreshToken.revoked || refreshToken.expiresAt < new Date()) {
            await this.logout(req, res);
            throw new HttpError(HttpStatus.UNAUTHORIZED, "Refresh token scaduto o revocato");
        }

        const userResult = await db.select().from(users).where(eq(users.id, refreshToken.userId));
        const user = userResult.at(0);

        if (!user) {
            await this.logout(req, res);
            throw new HttpError(HttpStatus.UNAUTHORIZED, "Nessun utente associato a questo token");
        }

        await db.update(refreshTokens)
            .set({ revoked: true })
            .where(eq(refreshTokens.token, refreshTokenValue));

        const newRefreshToken = generateRefreshToken();

        await db.insert(refreshTokens).values({
            userId: user.id,
            token: newRefreshToken,
            expiresAt: refreshToken.expiresAt,
            revoked: false
        });

        const remainingMs = refreshToken.expiresAt.getTime() - Date.now();

        res.cookie("refreshToken", newRefreshToken, {
            httpOnly: true,
            secure: cookieConfig.secure,
            path: "/api/auth",
            sameSite: cookieConfig.sameSite,
            maxAge: remainingMs
        });

        const accessToken = generateAccessToken({ sub: user.username, role: user.role });

        return { accessToken };
    }
}